File: base/src/main/java/me/shetj/base/base/BaseActivity.java
package me.shetj.base.base;

import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.Keep;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.util.AttributeSet;
import android.view.View;

import com.tbruyelle.rxpermissions2.RxPermissions;
import com.trello.rxlifecycle2.components.support.RxAppCompatActivity;

import org.simple.eventbus.EventBus;
import org.xutils.common.util.LogUtil;

import me.shetj.base.R;
import me.shetj.base.tools.app.ArmsUtils;
import me.shetj.base.tools.app.HideUtil;
import me.shetj.base.tools.json.EmptyUtils;
import me.shetj.base.view.LoadingDialog;

import static me.shetj.base.tools.app.ThirdViewUtil.convertAutoView;

/**
 * 基础类  view 层
 * @author shetj
 */
@Keep
public abstract class BaseActivity<T extends BasePresenter> extends RxAppCompatActivity implements IView {

    protected T mPresenter ;
    protected RxPermissions rxPermissions = null;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        //如果要使用eventbus请将此方法返回true
        if (useEventBus()) {
            //注册到事件主线
            EventBus.getDefault().register(this);
        }
        HideUtil.init(this);
        rxPermissions = new RxPermissions(this);
        startAnimation();

    }

    @Override
    protected void onResume() {
        super.onResume();
    }

    @Override
    protected void onPause() {
        super.onPause();
    }

    /**
     * 连接view
     */
    protected abstract void initView();

    /**
     * 连接数据
     */
    protected abstract void initData();

    @Override
    public View onCreateView(String name, Context context, AttributeSet attrs) {
        View view = convertAutoView(name, context, attrs);
        return view == null ? super.onCreateView(name, context, attrs) : view;
    }
    /**
     * 针对6.0动态请求权限问题
     * 判断是否允许此权限
     *
     * @param permissions  权限
     * @return hasPermission
     */
    protected boolean hasPermission(String... permissions) {
        for (String permission : permissions) {
            if (ContextCompat.checkSelfPermission(this, permission)
                    != PackageManager.PERMISSION_GRANTED) {
                return false;
            }
        }
        return true;
    }


    /**
     * 是否使用eventBus,默认为使用(true)，
     *
     * @return useEventBus
     */
    protected boolean useEventBus() {
        return true;
    }


    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (useEventBus()){
            //如果要使用eventbus请将此方法返回true
            EventBus.getDefault().unregister(this);
        }
        if (null != mPresenter) {
            mPresenter.onDestroy();
        }
    }

    @Override
    public void showLoading(String msg) {

       LoadingDialog.showLoading(this, msg, true);
    }
    @Override
    public void hideLoading() {
        LoadingDialog.hideLoading();
    }

    /**
     * 界面开始动画 (此处输入方法执行任务.)
     */
    protected void startAnimation() {
        overridePendingTransition(R.anim.push_right_in, R.anim.push_left_out);
    }

    /**
     * 界面回退动画 (此处输入方法执行任务.)
     */
    protected void endAnimation() {// 开始动画
        overridePendingTransition(R.anim.push_left_in, R.anim.push_right_out);
    }




    @Override
    public void finish() {// 设置回退动画
        super.finish();
    }

    /**
     * 返回
     */
    public void back() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            finishAfterTransition();
        } else {
            finish();
        }
    }




    @Override
    public void showMessage(@NonNull String message) {
        ArmsUtils.makeText(message);
    }

    @Override
    public void onBackPressed() {
        HideUtil.hideSoftKeyboard(getRxContext());
        super.onBackPressed();
        endAnimation();
        back();
    }

    @Override
    public RxAppCompatActivity getRxContext() {
        return this;
    }

    @SuppressWarnings("unchecked")
    @Override
    public void updateView(BaseMessage message){
        if (EmptyUtils.isNotEmpty(message)) {
            LogUtil.i(message.obj.toString());
        }
    }
}

File: base/src/main/java/me/shetj/base/base/BaseActivity.kt
package me.shetj.base.base


import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.view.View
import android.widget.TextView
import androidx.annotation.Keep
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.OnLifecycleEvent
import com.trello.rxlifecycle3.components.support.RxAppCompatActivity
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancelChildren
import me.shetj.base.R
import me.shetj.base.kt.toJson
import me.shetj.base.s
import me.shetj.base.tools.app.HideUtil
import me.shetj.base.tools.json.EmptyUtils
import org.simple.eventbus.EventBus
import timber.log.Timber
import kotlin.coroutines.CoroutineContext


/**
 * 基础类  view 层
 * @author shetj
 */
@Keep
abstract class BaseActivity<T : BasePresenter<*>> : RxAppCompatActivity(), IView , CoroutineScope,LifecycleObserver {
    protected val TAG = this.javaClass.simpleName
    private  var myHandler: Handler ?=null
    protected var mPresenter: T? = null

    override val rxContext: RxAppCompatActivity
        get() = this

    private val job = SupervisorJob()


    override val coroutineContext: CoroutineContext
        get() = Dispatchers.Main + job

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        startAnimation()
        lifecycle.addObserver(this)
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
    open fun onActivityCreate(){
        HideUtil.init(this)
        if (useEventBus()) {
            //注册到事件主线
            EventBus.getDefault().register(this)
        }
        findViewById<View>(R.id.toolbar_back)?.setOnClickListener { back() }
        initView()
        initData()
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    open fun onActivityDestroy(){
        if (useEventBus()) {
            //如果要使用eventbus请将此方法返回true
            EventBus.getDefault().unregister(this)
        }
        coroutineContext.cancelChildren()
        mPresenter?.onDestroy()
    }

    open fun setTitle(title:String){
        findViewById<TextView>(R.id.toolbar_title)?.apply {
            text = title
        }
    }
    /**
     * 连接view
     */
    protected abstract fun initView()
    /**
     * 连接数据
     */
    protected abstract fun initData()


    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        //true - 界面加载成功的时候
    }

    /**
     * 是否使用eventBus,默认为使用(true)，
     *
     * @return useEventBus
     */
    open fun useEventBus(): Boolean {
        return true
    }

    /**
     * 界面开始动画 (此处输入方法执行任务.)
     */
    open fun startAnimation() {}

    /**
     * 界面回退动画 (此处输入方法执行任务.)
     */
    open fun endAnimation() {}

    /**
     * 返回
     */
    open  fun back() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            finishAfterTransition()
        } else {
            finish()
        }
    }

    fun addDispose(disposable: Disposable) {
         mPresenter?.addDispose(disposable)
    }

    override fun onBackPressed() {
        HideUtil.hideSoftKeyboard(rxContext)
        super.onBackPressed()
        back()
    }

    override fun updateView(message: Message) {
        if (s.isDebug && EmptyUtils.isNotEmpty(message)) {
            Timber.i(message.toJson())
        }
    }
}
